src/js/ripple/keypairs.js

```diff
@@ -0,0 +1,380 @@
+  throw new Error('virtual method not implemented ');
+}
+
+function hasCachedProperty(obj, name, computer) {
+  var key = name + '__';
+  obj.prototype[name] = function() {
+    return this[key] !== undefined ? this[key] :
+           this[key] = computer.call(this);
+  };
+}
+
+/* --------------------------------- HELPERS -------------------------------- */
+
+function toGenericArray(typedArray) {
+  var generic = [];
+  Array.prototype.push.apply(generic, typedArray);
+  return generic;
+}
+
+function bytesToHex(bytes) {
+  return arrayToHex(bytes).toUpperCase();
+}
+
+/*
+
+Seed.from_json is too loose, and discriminates between seed and passphrase via
+the first letter of the passed in string. This requires explicitly declaring
+what format the seed is in.
+
+@private
+@param {Object} opts
+@param {String} [opts.base58] -
+@param {String} [opts.hex] -
+@param {String} [opts.passphrase] -
+@param {String} [opts.generic] -
+ */
+function parseSeed(opts) {
+  var seed = new Seed();
+
+  if (opts.passphrase) {
+    seed.parse_passphrase(opts.passphrase);
+  } else if (opts.base58) {
+    seed.parse_base58(opts.base58);
+  } else if (opts.hex) {
+    seed.parse_hex(opts.hex);
+  } else if (opts.generic) {
+    seed.parse_json(opts.generic);
+  } else {
+    throw new Error('generic, base58, hex, or passphrase must be supplied');
+  }
+
+  return seed;
+}
+
+/*
+@private
+@param {Seed} seed
+ */
+function deriveEdKeyPairSeed(seed) {
+  var hashed = sjcl.hash.sha512.hash(seed.to_bits());
+  var bits = sjcl.bitArray.bitSlice(hashed, 0, 256);
+  return new Uint8Array(sjcl.codec.bytes.fromBits(bits));
+}
+
+function findk256Key(bytes, discrim) {
+  var curve = sjcl.ecc.curves.k256;
+  var key;
+
+  for (var i = 0; i <= 0xFFFFFFFF; i++) {
+    // We hash the bytes to find a 256 bit number, looping until we are sure
+    // it is less than the order of the curve.
+    var hasher = new utils.Sha512().addBytes(bytes);
+    // If the optional discriminator index was passed in, update the hash.
+    if (discrim !== undefined) {
+      hasher.addU32(discrim);
+    }
+    hasher.addU32(i);
+    key = hasher.finish256BN();
+
+    /* eslint-disable max-len */
+
+    /*
+    # js
+    >>> console.log(curve.r.toString(16).toUpperCase())
+    0XFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141
+
+    # python
+    >>> order = b'FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141'
+    >>> import struct, hashlib, os
+    >>> randoms = lambda: (os.urandom(16) for i in range(2**12))
+    >>> h256 = lambda b, i: int(hashlib.sha512(b+struct.pack('>I', i)).hexdigest()[:64], 16)
+    >>> (2 ** 12) ** 2
+    16777216
+    >>> all(h256(r, i) < order for r in randoms() for i in range(2**12))
+    True
+
+    We basically always break here, but for the vanishingly rare case that we
+    don't we'll increment i and continue. :)
+
+    */
+
+    /* eslint-enable max-len */
+    if (key.greaterEquals(1) && curve.r.greaterEquals(key)) {
```

Comment:
There's a subtle logic bug here (also in the existing seed.js code). We want a value greater than zero, and less than (but not equal to) the order of the curve.  sjcl.bn doesn't provide us with the expressiveness to simply say `key < order`, so we must use the messier logic of `!(key >= order)`, which amounts to `!key.greaterEquals(curve.r)`